### ISSUE TITLE
Refactoring

### ISSUE THREAD

**Issue by TheHllm** [TGlas/tscript#55]
The code **badly** needs refactoring.
Some things that i propose need changing:

- not having a distribution folder / using ci to update a live version on e.g. an other github repositry
- proper build process (move to [typescript](https://stackoverflow.com/a/58042841/8055836))
- separating components into files (move to typescript)
- separating files into folders (for codemirror etc.)

But with 345Kb of js in a single file implementing any of these would likely require more than one person and a consensus on approaching this.

**Comment by TheHllm** [comment]
PoC @ https://github.com/TheHllm/tscript/tree/restructure.
Hello-world works.

**Comment by Jartreg** [comment]
**Yes!!**

It's definitely true that keeping everything in one file is not going to be sustainable. I'm totally supporting the idea, but I think there is more discussion and explanation necessary. For example, I don't know how many of the other contributors have worked with node/npm/typescript/bundlers/..., given that this project seems to have started without that.

TypeScript is something that I would totally love to see in this project, but I'm really not sure whether the other contributors would be comfortable with it. As a compromise, we could use TypeScript's type checker even in js files by using [JSDoc comments](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) and later enabling type checking by using [`// @ts-check`](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html).

Also, I agree that this shouldn't be done in just one pull request. I would suggest that we discuss and implement roughly the following steps one by one:

1. Get `distribution` outta here! These files should be on another branch like `gh-pages`.
2. Automate the creation of those files with [Actions](https://github.com/features/actions).
3. Switch the whole thing to using a bundler to build.
4. Use the bundler setup and slowly pull things out of the current js files one by one. While doing that, we can use the opportunity to make some restructurings, switch to modern ES syntax and annotate everything with JSDoc comments.

* (n - 1). Do something about the unit tests 
* n. Enable type checking with `// @ts-check`.

I deliberately didn't include any reasons/arguments here, but we shouldn't underestimate step 3. Currently you can start working on this project and see results without installing anything or using any build tool. Assuming that there might be people who don't know about Node, this step needs to be discussed in great detail.

Concerning your proof of concept, it's good that we can show _that_ it works, but I think we shouldn't use it as a reference because IMO it makes too many assumptions ahead of any discussions.

For now, I think we should wait for feedback from @TGlas to see what he thinks about any of this.

**Comment by Jartreg** [comment]
Btw, could you please correct the title to "Refactoring"? I will just hide this comment then...

**Comment by TGlas** [comment]
I do see your point in general, and I agree that the structure is not at all good. I think that it is clearly visible from the code structure that it was not planned from the start to end up like it did. When I started I was not very experienced with JS. Already the line count of some of the functions clearly tells you that they were evolving quite heavily; the expression parser is maybe a good example. ES6 is required, but its features are not used consequently. The "build system" (distribution folder, Makefile) is extremely old-school. It does the job, and admittedly, I did not aim for a proper solution. The main reason is that it worked for me, and TScript was never intended to be a community project. That said, I am very happy that you are willing to dedicate time into this!

The main reason why overhauling the structure was never necessary is because I don't have an anbitious vision of where the project should be going in the future. In terms of language features I like the current state, althoug one could of course discuss a few minor additions. I have a few things on my TODO list, but the major points there are unrelated to the project structure. For example, I would like to improve the documentation, and I'd like to create a tutorial, along the lines of the first half of the lecture. Without a need to develop the project much further, refactoring did not make it to the top of my TODO list.

Another aspect is that currently I don't have the time to do this. Implementing the language core took a couple of weeks (mostly evenings), and implementing all current features plus writing the documentation took several month. That was more than two years ago. In contrast to that time, I am now nearly completely booked with ongoing projects. Therefore, someone else would need to take the lead in this.

Personally, I never worked with Typescript. I checked the language carefully a few years back, and I liked it very much. In terms of features, it is nearly perfect. The main downside I see is that it does not execute natively in the browser. That may somewhat hinder debugging (although I read that there are good solutions), and more importantly: it may make it more difficult to incorporate the language into other projects. Currently, the main use case is its embedding into the e-learning system of the course. On the client side, that happens in the browser. On the server side I use only the core file found in `source/tscript.js` with node, together with a small wrapper script. This requirement is the main reason for the optional hooks in the interpreter, which are also used by the recently added audio support. In the foreseeable future, the same thing needs to work in an electronic exam system. So my main requirement for a possible transition to a different language is to keep things easily inter-operable with such systems, which will likely be written in plain JS on the client side. I have no idea how easy of difficult it will be to get this right.
